### ISSUE TITLE
regression in 2.7.0: false positives use-a-generator

### ISSUE THREAD

**Issue by sam-s** [PyCQA/pylint#4133]
### Steps to reproduce

The following code generates diagnostics in 2.7.0 but not in 2.6.0:
```python
def isclose(a, b, tolerances=None):
    """Call math.isclose() recursively.
    https://stackoverflow.com/q/55497837/850781"""
    if a == b:
        return True
    tolerances = tolerances or {}
    if isinstance(a,(float,int)) and isinstance(b,(float,int)):
        return math.isclose(a, b, **tolerances)
    if ((isinstance(a, tuple) and isinstance(b, tuple)) or
        (isinstance(a, list) and isinstance(b, list))) and len(a) == len(b):
        return all([isclose(x,y,tolerances=tolerances) for x,y in zip(a,b)])
    if isinstance(a, dict) and isinstance(b, dict):
        return a.keys() == b.keys() and all([
            isclose(v,b[k],tolerances=tolerances) for k,v in a.items()])
    return False
```    

### Current behavior
Diagnostics:
```
[R1729(use-a-generator), isclose] Use a generator instead 'all(isclose(x, y, tolerances=tolerances) for (x, y) in zip(a, b))'
[R1729(use-a-generator), isclose] Use a generator instead 'all(isclose(v, b[k], tolerances=tolerances) for (k, v) in a.items())'
```

### Expected behavior
I don't see how to use a generator here.
(In the first case, I suppose I could write `all(map((lambda x,y:isclose(x,y,tolerances=tolerances)),a,b))` but I don't see how that is better than the code I have now).

### pylint --version output
```
pylint 2.7.0
astroid 2.5
Python 3.9.2 (default, Feb 19 2021, 17:09:53) 
[Clang 12.0.0 (clang-1200.0.32.29)]
```

**Comment by sam-s** [comment]
cf https://github.com/PyCQA/pylint/issues/4132

**Comment by ikraduya** [comment]
Please refer to this #3309 
Especially at [this](https://github.com/PyCQA/pylint/pull/3309#discussion_r577160339) discussion

TL;DR: removing "[ ]" inside `any` or `all` function will give better performance in most cases, since you can cut the execution tree and exit directly at the first element.
